File: apps/cptr/cptr/services/mcp_diagnostics.py
"""Bounded allowlisted MCP diagnostics and backend system telemetry state."""

from __future__ import annotations

import asyncio
import math
import os
from collections import deque
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from cptr.services.mcp_pricing import project_usage_cost
from cptr.utils.redaction import redact_external_text

LatencyEdge = Literal[
    "client-mcp-connector",
    "mcp-connector-cptr-mcp",
    "cptr-mcp-cptr-backend",
]
LatencyMetric = Literal[
    "observed_request_time",
    "adapter_handoff",
    "backend_api_rtt",
]
FailureStage = Literal[
    "client_transport",
    "mcp_connector",
    "cptr_mcp",
    "cptr_backend",
    "activity_delivery",
    "traffic_delivery",
]


class McpLatencySample(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["latency"] = "latency"
    version: Literal[1] = 1
    event_id: str = Field(min_length=8, max_length=128)
    timestamp_ms: int = Field(ge=0)
    request_id: str | None = Field(default=None, max_length=128)
    correlation_id: str | None = Field(default=None, max_length=128)
    edge_id: LatencyEdge
    metric_type: LatencyMetric
    duration_ms: int = Field(ge=0, le=86_400_000)
    status: Literal["ok", "error"] = "ok"


class McpUsageDiagnostic(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["usage"] = "usage"
    version: Literal[1] = 1
    event_id: str = Field(min_length=8, max_length=128)
    timestamp_ms: int = Field(ge=0)
    request_id: str | None = Field(default=None, max_length=128)
    correlation_id: str | None = Field(default=None, max_length=128)
    session_id: str | None = Field(default=None, max_length=128)
    client_id: str = Field(min_length=1, max_length=128)
    model_reported: str | None = Field(default=None, max_length=120)
    model_canonical: str | None = Field(default=None, max_length=64)
    model_source: Literal["self_reported", "unavailable"]
    tool_name: str = Field(min_length=1, max_length=256)
    input_tokens_estimated: int = Field(ge=0, le=100_000_000)
    output_tokens_estimated: int = Field(ge=0, le=100_000_000)
    cached_input_tokens_estimated: None = None
    estimator_method: str = Field(min_length=1, max_length=160)
    estimator_exact_for_model: bool
    status: Literal["complete", "error"]


class McpFailureDiagnostic(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["failure"] = "failure"
    version: Literal[1] = 1
    diagnostic_id: str = Field(min_length=8, max_length=128)
    request_id: str | None = Field(default=None, max_length=128)
    correlation_id: str | None = Field(default=None, max_length=128)
    session_id: str | None = Field(default=None, max_length=128)
    client_id: str = Field(min_length=1, max_length=128)
    method: str | None = Field(default=None, max_length=128)
    tool_name: str | None = Field(default=None, max_length=256)
    stage: FailureStage
    error_code: str = Field(min_length=1, max_length=64)
    http_status: int | None = Field(default=None, ge=100, le=599)
    retryable: bool | None = None
    started_at_ms: int | None = Field(default=None, ge=0)
    completed_at_ms: int = Field(ge=0)
    duration_ms: int | None = Field(default=None, ge=0, le=86_400_000)
    request_bytes: int | None = Field(default=None, ge=0, le=100_000_000)
    response_bytes: int | None = Field(default=None, ge=0, le=100_000_000)
    summary: str = Field(min_length=1, max_length=500)


class McpGpuMetrics(BaseModel):
    model_config = ConfigDict(extra="forbid")

    index: int = Field(ge=0, le=64)
    name: str = Field(min_length=1, max_length=120)
    utilization_percent: float = Field(ge=0, le=100)
    memory_used_bytes: int = Field(ge=0)
    memory_total_bytes: int = Field(ge=0)
    temperature_c: float | None = Field(default=None, ge=-50, le=150)


class McpProcessMetrics(BaseModel):
    model_config = ConfigDict(extra="forbid")

    pid: int = Field(ge=0)
    cpu_percent: float | None = Field(default=None, ge=0)
    memory_percent: float | None = Field(default=None, ge=0, le=100)
    name: str = Field(min_length=1, max_length=160)


class McpBackendMetricsSample(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["system"] = "system"
    version: Literal[1] = 1
    timestamp_ms: int = Field(ge=0)
    cpu_usage_percent: float | None = Field(default=None, ge=0, le=100)
    cpu_count: int = Field(ge=0, le=4096)
    load_avg: list[float] = Field(default_factory=list, max_length=3)
    memory_total_bytes: int | None = Field(default=None, ge=0)
    memory_available_bytes: int | None = Field(default=None, ge=0)
    disk_total_bytes: int | None = Field(default=None, ge=0)
    disk_used_bytes: int | None = Field(default=None, ge=0)
    disk_free_bytes: int | None = Field(default=None, ge=0)
    disk_read_bytes_per_s: float | None = Field(default=None, ge=0)
    disk_write_bytes_per_s: float | None = Field(default=None, ge=0)
    disk_read_ops_per_s: float | None = Field(default=None, ge=0)
    disk_write_ops_per_s: float | None = Field(default=None, ge=0)
    network_rx_bytes_per_s: float | None = Field(default=None, ge=0)
    network_tx_bytes_per_s: float | None = Field(default=None, ge=0)
    uptime_seconds: int | None = Field(default=None, ge=0)
    gpu_status: Literal["available", "unavailable", "error"] = "unavailable"
    gpus: list[McpGpuMetrics] = Field(default_factory=list, max_length=16)
    cptr_process: McpProcessMetrics | None = None
    processes: list[McpProcessMetrics] = Field(default_factory=list, max_length=10)


class McpDiagnosticsBatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    events: list[McpLatencySample | McpFailureDiagnostic | McpUsageDiagnostic] = Field(
        min_length=1, max_length=100
    )


def _bounded_env_int(name: str, default: int, minimum: int, maximum: int) -> int:
    raw = os.getenv(name)
    if raw is None:
        return default
    try:
        value = int(raw)
    except (TypeError, ValueError):
        return default
    return max(minimum, min(value, maximum))


def _nearest_rank(values: list[int], percentile: float) -> int:
    ordered = sorted(values)
    if not ordered:
        return 0
    rank = max(1, math.ceil(percentile * len(ordered)))
    return ordered[min(len(ordered), rank) - 1]


class McpDiagnosticsStore:
    """Bounded in-memory diagnostics with non-blocking subscriber fan-out."""

    def __init__(
        self,
        *,
        max_latency_samples_per_edge: int = 120,
        max_failures: int = 250,
        max_system_samples: int = 60,
        max_usage: int = 500,
        subscriber_queue_size: int = 64,
        observed_degraded_ms: int | None = None,
        handoff_degraded_ms: int | None = None,
        backend_rtt_degraded_ms: int | None = None,
    ) -> None:
        self.max_latency_samples_per_edge = max(1, int(max_latency_samples_per_edge))
        self.max_failures = max(1, int(max_failures))
        self.max_system_samples = max(1, int(max_system_samples))
        self.max_usage = max(1, int(max_usage))
        self.subscriber_queue_size = max(1, int(subscriber_queue_size))
        self._thresholds: dict[str, int] = {
            "observed_request_time": observed_degraded_ms
            or _bounded_env_int("CPTR_MCP_DIAGNOSTICS_OBSERVED_DEGRADED_MS", 5000, 1, 86_400_000),
            "adapter_handoff": handoff_degraded_ms
            or _bounded_env_int("CPTR_MCP_DIAGNOSTICS_HANDOFF_DEGRADED_MS", 100, 1, 86_400_000),
            "backend_api_rtt": backend_rtt_degraded_ms
            or _bounded_env_int(
                "CPTR_MCP_DIAGNOSTICS_BACKEND_RTT_DEGRADED_MS", 1000, 1, 86_400_000
            ),
        }
        self._latency: dict[str, deque[dict[str, object]]] = {}
        self._failures: deque[dict[str, object]] = deque(maxlen=self.max_failures)
        self._system: deque[dict[str, object]] = deque(maxlen=self.max_system_samples)
        self._usage: deque[dict[str, object]] = deque(maxlen=self.max_usage)
        # Keep the event history bounded while preserving process-lifetime cumulative
        # counters. This makes the dashboard totals genuinely cumulative without
        # allowing telemetry memory to grow with request volume.
        self._usage_total_input_tokens = 0
        self._usage_total_output_tokens = 0
        self._usage_total_simulated_cost = Decimal("0")
        self._usage_priced_events = 0
        self._usage_stale_events = 0
        self._usage_unpriced_events = 0
        self._usage_by_model: dict[str, dict[str, object]] = {}
        self._seen_ids: deque[str] = deque()
        self._seen_id_set: set[str] = set()
        self._dedupe_capacity = max(
            64,
            (self.max_latency_samples_per_edge * 3 + self.max_failures + self.max_usage) * 4,
        )
        self._subscribers: set[asyncio.Queue[dict[str, object]]] = set()
        self._lock = asyncio.Lock()
        self._sequence = 0
        self._slow_subscriber_drops = 0

    def subscribe(self) -> asyncio.Queue[dict[str, object]]:
        queue: asyncio.Queue[dict[str, object]] = asyncio.Queue(maxsize=self.subscriber_queue_size)
        self._subscribers.add(queue)
        return queue

    def unsubscribe(self, queue: asyncio.Queue[dict[str, object]]) -> None:
        self._subscribers.discard(queue)

    async def ingest(
        self, events: list[McpLatencySample | McpFailureDiagnostic | McpUsageDiagnostic]
    ) -> dict[str, int]:
        accepted = 0
        duplicates = 0
        dropped = 0
        async with self._lock:
            for event in events:
                event_id = (
                    event.event_id
                    if isinstance(event, (McpLatencySample, McpUsageDiagnostic))
                    else event.diagnostic_id
                )
                if event_id in self._seen_id_set:
                    duplicates += 1
                    continue
                self._remember_id(event_id)
                self._sequence += 1

                if isinstance(event, McpLatencySample):
                    projected = event.model_dump(mode="json")
                    bucket = self._latency.setdefault(
                        event.edge_id,
                        deque(maxlen=self.max_latency_samples_per_edge),
                    )
                    bucket.append(projected)
                elif isinstance(event, McpUsageDiagnostic):
                    projected = project_usage_cost(event)
                    self._accumulate_usage(projected)
                    self._usage.append(projected)
                else:
                    safe_summary = redact_external_text(event.summary).strip()[:500]
                    if not safe_summary:
                        safe_summary = "MCP request failed."
                    projected = event.model_copy(update={"summary": safe_summary}).model_dump(
                        mode="json"
                    )
                    self._failures.append(projected)

                projected = {**projected, "ingestion_sequence": self._sequence}
                self._fan_out(projected)
                accepted += 1
        return {"accepted": accepted, "duplicates": duplicates, "dropped": dropped}

    async def record_system_sample(self, sample: McpBackendMetricsSample) -> None:
        async with self._lock:
            self._sequence += 1
            projected = sample.model_dump(mode="json")
            self._system.append(projected)
            self._fan_out({**projected, "ingestion_sequence": self._sequence})

    async def snapshot(self) -> dict[str, object]:
        async with self._lock:
            latency = {
                edge_id: self._latency_aggregate(samples)
                for edge_id, samples in sorted(self._latency.items())
                if samples
            }
            return {
                "version": 1,
                "sequence": self._sequence,
                "latency": latency,
                "failures": list(self._failures),
                "system": list(self._system),
                "usage": list(self._usage),
                "current_model": dict(self._usage[-1]) if self._usage else None,
                "usage_totals": self._usage_totals(),
                "stream_health": {
                    "subscriber_count": len(self._subscribers),
                    "slow_subscriber_drops": self._slow_subscriber_drops,
                    "latency_sample_capacity_per_edge": self.max_latency_samples_per_edge,
                    "failure_capacity": self.max_failures,
                    "system_sample_capacity": self.max_system_samples,
                    "usage_capacity": self.max_usage,
                    "subscriber_queue_capacity": self.subscriber_queue_size,
                },
            }

    def _accumulate_usage(self, item: dict[str, object]) -> None:
        item_input = int(item.get("input_tokens_estimated") or 0)
        item_output = int(item.get("output_tokens_estimated") or 0)
        self._usage_total_input_tokens += item_input
        self._usage_total_output_tokens += item_output

        status = str(item.get("pricing_status") or "unknown_model")
        if status == "current":
            self._usage_priced_events += 1
        elif status == "stale":
            self._usage_stale_events += 1
        else:
            self._usage_unpriced_events += 1

        raw_cost = item.get("simulated_cost_usd")
        item_cost = Decimal(str(raw_cost)) if raw_cost is not None else Decimal("0")
        self._usage_total_simulated_cost += item_cost

        model_id = item.get("model_canonical")
        if not isinstance(model_id, str) or not model_id:
            return
        group = self._usage_by_model.setdefault(
            model_id,
            {
                "events": 0,
                "input_tokens_estimated": 0,
                "output_tokens_estimated": 0,
                "total_tokens_estimated": 0,
                "simulated_cost_usd": Decimal("0"),
            },
        )
        group["events"] = int(group["events"]) + 1
        group["input_tokens_estimated"] = int(group["input_tokens_estimated"]) + item_input
        group["output_tokens_estimated"] = int(group["output_tokens_estimated"]) + item_output
        group["total_tokens_estimated"] = (
            int(group["total_tokens_estimated"]) + item_input + item_output
        )
        group["simulated_cost_usd"] = Decimal(str(group["simulated_cost_usd"])) + item_cost

    def _usage_totals(self) -> dict[str, object]:
        serialized_by_model = {
            model_id: {
                **values,
                "simulated_cost_usd": format(Decimal(str(values["simulated_cost_usd"])), "f"),
            }
            for model_id, values in sorted(self._usage_by_model.items())
        }
        return {
            "input_tokens_estimated": self._usage_total_input_tokens,
            "output_tokens_estimated": self._usage_total_output_tokens,
            "total_tokens_estimated": self._usage_total_input_tokens
            + self._usage_total_output_tokens,
            "simulated_cost_usd": format(self._usage_total_simulated_cost, "f"),
            "priced_events": self._usage_priced_events,
            "stale_events": self._usage_stale_events,
            "unpriced_events": self._usage_unpriced_events,
            "by_model": serialized_by_model,
        }

    def _latency_aggregate(self, samples: deque[dict[str, object]]) -> dict[str, object]:
        values = [int(sample["duration_ms"]) for sample in samples]
        latest = samples[-1]
        p95 = _nearest_rank(values, 0.95)
        metric_type = str(latest["metric_type"])
        latest_status = str(latest["status"])
        health = (
            "error"
            if latest_status == "error"
            else ("degraded" if p95 >= self._thresholds.get(metric_type, 86_400_000) else "healthy")
        )
        return {
            "metric_type": metric_type,
            "latest_ms": values[-1],
            "average_ms": sum(values) / len(values),
            "p50_ms": _nearest_rank(values, 0.50),
            "p95_ms": p95,
            "max_ms": max(values),
            "sample_count": len(values),
            "last_updated_ms": int(latest["timestamp_ms"]),
            "latest_status": latest_status,
            "health": health,
        }

    def _remember_id(self, event_id: str) -> None:
        self._seen_ids.append(event_id)
        self._seen_id_set.add(event_id)
        while len(self._seen_ids) > self._dedupe_capacity:
            expired = self._seen_ids.popleft()
            self._seen_id_set.discard(expired)

    def _fan_out(self, projected: dict[str, object]) -> None:
        for queue in tuple(self._subscribers):
            try:
                queue.put_nowait(projected)
                continue
            except asyncio.QueueFull:
                pass
            try:
                queue.get_nowait()
                self._slow_subscriber_drops += 1
            except asyncio.QueueEmpty:
                pass
            try:
                queue.put_nowait(projected)
            except asyncio.QueueFull:
                self._slow_subscriber_drops += 1


mcp_diagnostics_store = McpDiagnosticsStore(
    max_latency_samples_per_edge=_bounded_env_int(
        "CPTR_MCP_DIAGNOSTICS_LATENCY_SAMPLES_PER_EDGE", 120, 10, 2000
    ),
    max_failures=_bounded_env_int("CPTR_MCP_DIAGNOSTICS_MAX_FAILURES", 250, 10, 5000),
    max_system_samples=_bounded_env_int("CPTR_MCP_DIAGNOSTICS_MAX_SYSTEM_SAMPLES", 60, 10, 600),
    max_usage=_bounded_env_int("CPTR_MCP_DIAGNOSTICS_MAX_USAGE", 500, 10, 10_000),
    subscriber_queue_size=_bounded_env_int("CPTR_MCP_DIAGNOSTICS_SUBSCRIBER_QUEUE", 64, 8, 512),
)

File: apps/mcp/server/mcp-diagnostics.ts
import { randomUUID } from "node:crypto";

export type McpLatencyEdge =
  | "client-mcp-connector"
  | "mcp-connector-cptr-mcp"
  | "cptr-mcp-cptr-backend";

export type McpLatencyMetric = "observed_request_time" | "adapter_handoff" | "backend_api_rtt";

export type McpFailureStage =
  | "client_transport"
  | "mcp_connector"
  | "cptr_mcp"
  | "cptr_backend"
  | "activity_delivery"
  | "traffic_delivery";

export type McpLatencyDiagnostic = {
  kind: "latency";
  version: 1;
  event_id: string;
  timestamp_ms: number;
  request_id: string | null;
  correlation_id: string | null;
  edge_id: McpLatencyEdge;
  metric_type: McpLatencyMetric;
  duration_ms: number;
  status: "ok" | "error";
};

export type McpUsageDiagnostic = {
  kind: "usage";
  version: 1;
  event_id: string;
  timestamp_ms: number;
  request_id: string | null;
  correlation_id: string | null;
  session_id: string | null;
  client_id: string;
  model_reported: string | null;
  model_canonical: string | null;
  model_source: "self_reported" | "unavailable";
  tool_name: string;
  input_tokens_estimated: number;
  output_tokens_estimated: number;
  cached_input_tokens_estimated: null;
  estimator_method: string;
  estimator_exact_for_model: boolean;
  status: "complete" | "error";
};

export type McpFailureDiagnostic = {
  kind: "failure";
  version: 1;
  diagnostic_id: string;
  request_id: string | null;
  correlation_id: string | null;
  session_id: string | null;
  client_id: string;
  method: string | null;
  tool_name: string | null;
  stage: McpFailureStage;
  error_code: string;
  http_status: number | null;
  retryable: boolean | null;
  started_at_ms: number | null;
  completed_at_ms: number;
  duration_ms: number | null;
  request_bytes: number | null;
  response_bytes: number | null;
  summary: string;
};

export type McpDiagnosticEvent = McpLatencyDiagnostic | McpFailureDiagnostic | McpUsageDiagnostic;

export type McpDiagnosticsEmitterOptions = {
  deliver: (events: McpDiagnosticEvent[]) => Promise<void>;
  env?: Record<string, string | undefined>;
  batchSize?: number;
  flushMs?: number;
  maxQueue?: number;
};

function boundedInt(raw: string | undefined, fallback: number, minimum: number, maximum: number): number {
  const parsed = raw === undefined ? Number.NaN : Number.parseInt(raw, 10);
  if (!Number.isFinite(parsed)) return fallback;
  return Math.max(minimum, Math.min(maximum, parsed));
}

function boundedOverride(value: number | undefined, fallback: number, minimum: number, maximum: number): number {
  if (value === undefined || !Number.isFinite(value)) return fallback;
  return Math.max(minimum, Math.min(maximum, Math.round(value)));
}

function boundedText(value: unknown, maximum: number): string | null {
  if (typeof value !== "string") return null;
  const trimmed = value.trim().replace(/[\u0000-\u001f\u007f]/g, " ").replace(/\s+/g, " ");
  return trimmed ? trimmed.slice(0, maximum) : null;
}

function boundedNumber(value: number | null | undefined, maximum: number): number | null {
  if (typeof value !== "number" || !Number.isFinite(value)) return null;
  return Math.max(0, Math.min(maximum, Math.round(value)));
}

function boundedHttpStatus(value: number | null | undefined): number | null {
  if (typeof value !== "number" || !Number.isFinite(value)) return null;
  const rounded = Math.round(value);
  return rounded >= 100 && rounded <= 599 ? rounded : null;
}

export function sanitizeDiagnosticSummary(value: unknown): string {
  const text = typeof value === "string" ? value : "MCP request failed.";
  const redacted = text
    .replace(/\bBearer\s+[^\s,;]+/gi, "Bearer [REDACTED]")
    .replace(/(?:^|[\s:(])\/(?:[^\s/:]+\/)*[^\s/:]+/g, (match) => {
      const prefix = /^[\s:(]/.test(match) ? match[0] : "";
      return `${prefix}<redacted-path>`;
    })
    .replace(/[A-Za-z]:\\(?:[^\\\s]+\\)*[^\\\s]+/g, "<redacted-path>")
    .replace(/[\u0000-\u001f\u007f]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
  return (redacted || "MCP request failed.").slice(0, 500);
}

function copyLatency(event: McpLatencyDiagnostic): McpLatencyDiagnostic {
  return {
    kind: "latency",
    version: 1,
    event_id: boundedText(event.event_id, 128) ?? randomUUID(),
    timestamp_ms: boundedNumber(event.timestamp_ms, Number.MAX_SAFE_INTEGER) ?? Date.now(),
    request_id: boundedText(event.request_id, 128),
    correlation_id: boundedText(event.correlation_id, 128),
    edge_id: event.edge_id,
    metric_type: event.metric_type,
    duration_ms: boundedNumber(event.duration_ms, 86_400_000) ?? 0,
    status: event.status,
  };
}

function copyUsage(event: McpUsageDiagnostic): McpUsageDiagnostic {
  return {
    kind: "usage",
    version: 1,
    event_id: boundedText(event.event_id, 128) ?? randomUUID(),
    timestamp_ms: boundedNumber(event.timestamp_ms, Number.MAX_SAFE_INTEGER) ?? Date.now(),
    request_id: boundedText(event.request_id, 128),
    correlation_id: boundedText(event.correlation_id, 128),
    session_id: boundedText(event.session_id, 128),
    client_id: (boundedText(event.client_id, 128) ?? "chatgpt").toLowerCase(),
    model_reported: boundedText(event.model_reported, 120),
    model_canonical: boundedText(event.model_canonical, 64),
    model_source: event.model_source === "self_reported" ? "self_reported" : "unavailable",
    tool_name: boundedText(event.tool_name, 256) ?? "unknown-tool",
    input_tokens_estimated: boundedNumber(event.input_tokens_estimated, 100_000_000) ?? 0,
    output_tokens_estimated: boundedNumber(event.output_tokens_estimated, 100_000_000) ?? 0,
    cached_input_tokens_estimated: null,
    estimator_method: boundedText(event.estimator_method, 160) ?? "unknown",
    estimator_exact_for_model: event.estimator_exact_for_model === true,
    status: event.status === "error" ? "error" : "complete",
  };
}

function copyFailure(event: McpFailureDiagnostic): McpFailureDiagnostic {
  return {
    kind: "failure",
    version: 1,
    diagnostic_id: boundedText(event.diagnostic_id, 128) ?? randomUUID(),
    request_id: boundedText(event.request_id, 128),
    correlation_id: boundedText(event.correlation_id, 128),
    session_id: boundedText(event.session_id, 128),
    client_id: (boundedText(event.client_id, 128) ?? "chatgpt").toLowerCase(),
    method: boundedText(event.method, 128),
    tool_name: boundedText(event.tool_name, 256),
    stage: event.stage,
    error_code: boundedText(event.error_code, 64) ?? "unknown_error",
    http_status: boundedHttpStatus(event.http_status),
    retryable: typeof event.retryable === "boolean" ? event.retryable : null,
    started_at_ms: boundedNumber(event.started_at_ms, Number.MAX_SAFE_INTEGER),
    completed_at_ms: boundedNumber(event.completed_at_ms, Number.MAX_SAFE_INTEGER) ?? Date.now(),
    duration_ms: boundedNumber(event.duration_ms, 86_400_000),
    request_bytes: boundedNumber(event.request_bytes, 100_000_000),
    response_bytes: boundedNumber(event.response_bytes, 100_000_000),
    summary: sanitizeDiagnosticSummary(event.summary),
  };
}

function copyEvent(event: McpDiagnosticEvent): McpDiagnosticEvent {
  if (event.kind === "latency") return copyLatency(event);
  if (event.kind === "usage") return copyUsage(event);
  return copyFailure(event);
}

export class McpDiagnosticsEmitter {
  private readonly deliver: (events: McpDiagnosticEvent[]) => Promise<void>;
  private readonly batchSize: number;
  private readonly flushMs: number;
  private readonly maxQueue: number;
  private queue: McpDiagnosticEvent[] = [];
  private dropped = 0;
  private timer: ReturnType<typeof setTimeout> | null = null;
  private deliveryPromise: Promise<void> | null = null;
  private closed = false;

  constructor(options: McpDiagnosticsEmitterOptions) {
    const env = options.env ?? process.env;
    const envBatchSize = boundedInt(env.CPTR_MCP_DIAGNOSTICS_PLUGIN_BATCH_SIZE, 20, 1, 100);
    const envFlushMs = boundedInt(env.CPTR_MCP_DIAGNOSTICS_PLUGIN_FLUSH_MS, 250, 25, 10_000);
    const envMaxQueue = boundedInt(env.CPTR_MCP_DIAGNOSTICS_PLUGIN_MAX_QUEUE, 500, 10, 5_000);
    this.deliver = options.deliver;
    this.batchSize = boundedOverride(options.batchSize, envBatchSize, 1, 100);
    this.flushMs = boundedOverride(options.flushMs, envFlushMs, 1, 10_000);
    this.maxQueue = boundedOverride(options.maxQueue, envMaxQueue, 1, 5_000);
  }

  stats(): { queued: number; dropped: number; delivering: boolean } {
    return { queued: this.queue.length, dropped: this.dropped, delivering: this.deliveryPromise !== null };
  }

  latency(
    input: Omit<McpLatencyDiagnostic, "kind" | "version" | "event_id" | "timestamp_ms">,
  ): void {
    this.emit({
      kind: "latency",
      version: 1,
      event_id: randomUUID(),
      timestamp_ms: Date.now(),
      ...input,
    });
  }

  failure(
    input: Omit<McpFailureDiagnostic, "kind" | "version" | "diagnostic_id" | "completed_at_ms">,
  ): void {
    this.emit({
      kind: "failure",
      version: 1,
      diagnostic_id: randomUUID(),
      completed_at_ms: Date.now(),
      ...input,
    });
  }

  usage(
    input: Omit<McpUsageDiagnostic, "kind" | "version" | "event_id" | "timestamp_ms">,
  ): void {
    this.emit({
      kind: "usage",
      version: 1,
      event_id: randomUUID(),
      timestamp_ms: Date.now(),
      ...input,
    });
  }

  async flush(): Promise<void> {
    if (this.deliveryPromise) return this.deliveryPromise;
    this.clearTimer();
    if (this.queue.length === 0) return;
    this.deliveryPromise = this.drain();
    try {
      await this.deliveryPromise;
    } finally {
      this.deliveryPromise = null;
      if (!this.closed && this.queue.length > 0) this.schedule();
    }
  }

  async close(): Promise<void> {
    this.closed = true;
    this.clearTimer();
    if (this.deliveryPromise) await this.deliveryPromise;
    if (this.queue.length > 0) {
      this.deliveryPromise = this.drain();
      try {
        await this.deliveryPromise;
      } finally {
        this.deliveryPromise = null;
      }
    }
  }

  private emit(event: McpDiagnosticEvent): void {
    if (this.closed) return;
    if (this.queue.length >= this.maxQueue) {
      this.queue.shift();
      this.dropped += 1;
    }
    this.queue.push(copyEvent(event));
    this.schedule();
  }

  private schedule(): void {
    if (this.timer || this.deliveryPromise || this.closed) return;
    this.timer = setTimeout(() => {
      this.timer = null;
      void this.flush();
    }, this.flushMs);
    this.timer.unref?.();
  }

  private clearTimer(): void {
    if (this.timer) clearTimeout(this.timer);
    this.timer = null;
  }

  private async drain(): Promise<void> {
    while (this.queue.length > 0) {
      const batch = this.queue.splice(0, this.batchSize).map(copyEvent);
      try {
        await this.deliver(batch);
      } catch {
        this.dropped += batch.length;
      }
    }
  }
}
